src/utils/task_stats_alloc/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![feature(allocator_api)]
+#![feature(lint_reasons)]
+#![feature(atomic_mut_ptr)]
+
+use std::alloc::{GlobalAlloc, Layout, System};
+use std::future::Future;
+use std::sync::atomic::{fence, AtomicUsize, Ordering};
+use std::time::Duration;
+
+use tokio::task_local;
+
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug)]
+pub struct TaskLocalBytesAllocated(Option<&'static AtomicUsize>);
+
+impl Default for TaskLocalBytesAllocated {
+    fn default() -> Self {
+        Self(Some(Box::leak(Box::new_in(0.into(), System))))
+    }
+}
+
+impl TaskLocalBytesAllocated {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create an invalid counter.
+    pub const fn invalid() -> Self {
+        Self(None)
+    }
+
+    /// Adds to the current counter.
+    #[inline(always)]
+    fn add(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            bytes.fetch_add(val, Ordering::Relaxed);
+        }
+    }
+
+    /// Adds to the current counter without validity check.
+    ///
+    /// # Safety
+    /// The caller must ensure that `self` is valid.
+    #[inline(always)]
+    unsafe fn add_unchecked(&self, val: usize) {
+        self.0.unwrap_unchecked().fetch_add(val, Ordering::Relaxed);
+    }
+
+    /// Subtracts from the counter value, and `drop` the counter while the count reaches zero.
+    #[inline(always)]
+    fn sub(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            // Use Relax as we only need to make it atomic.
```

Comment:
A little suggestion for the comment:
```suggestion
            // Use `Relaxed` order as we don't need to sync read/write with other memory addresses.
            // Accesses to the counter itself are serialized by atomic operations.
```